src/odemis/driver/technolution.py

```diff
@@ -0,0 +1,1323 @@
+        resolution = mppc.cellCompleteResolution.value[0]
+
+        # Check if the descanner clockperiod is still a multiple of de scanner clock period, otherwise raise error.
+        if not almost_equal(descanner.clockPeriod.value % scanner.clockPeriod.value, 0):
+            logging.error("Descanner and/or scanner clock period changed which means the descanner period is no "
+                          "longer a whole multiple of scanner clock periods. This means that calibration is no "
+                          "longer accurate.")
+            raise ValueError("Descanner clock period is no longer a whole multiple of the scanner clock period.")
+
+        # Creation of setpoints
+        flyback_time = descanner.physicalFlybackTime
+        remainder_scanning_time = (dwell_time * resolution) % descanner.clockPeriod.value
+        if remainder_scanning_time is not 0:
+            # Add adjusted flyback time if there is a remainder of scanning time by adding one setpoint to ensure the
+            # total scanning period is equal to a whole number of descan clock periods.
+            flyback_time = flyback_time + descanner.clockPeriod.value
+
+        # Period of the calibration signal in seconds
+        calibration_period = (dwell_time * resolution) + flyback_time
+
+        # Support a calibration freq from 125 to 25000 Hz
+        if not scanner.dwellTime.range[0] < calibration_period < 0.008:
+            logging.error("Cannot perform calibration using a calibration frequency lower than 125 or 5000. With the "
+                          "given parameters the required calibration frequency would be %s." % (1/calibration_period))
+            raise ValueError("Calibration of given values requires a calibration frequency which is out of range.")
+        calibration_frequency = 1/calibration_period
+
+        # One period with points at a sampling period of the descanner clock period
+        time_points_descanner = numpy.arange(0, calibration_period, descanner.clockPeriod.value)
+        x_descan_setpoints = descanner.scanGain.value[0] * numpy.sin(2 * math.pi * calibration_frequency * time_points_descanner)
+        y_descan_setpoints = descanner.scanGain.value[1] * signal.sawtooth(2 * math.pi * calibration_frequency *
+                                                                           time_points_descanner)
+
+        # Scan sampling frequency is 9000 times the calibration frequency, meaning 9000 setpoints per period.
+        time_points_scanner = numpy.linspace(0, calibration_period, 9000)
+        x_scan_setpoints = scanner.scanGain.value[0] * numpy.sin(2 * math.pi * calibration_frequency * time_points_scanner)
+        y_scan_setpoints = scanner.scanGain.value[1] * signal.sawtooth(2 * math.pi * calibration_frequency *
+                                                                       time_points_scanner)
+
+        calibration_data = CalibrationLoopParameters(descan_rotation,
```

Comment:
Reading the code again, I am wondering if it is not cleaner and more explicit to check here, that the number of setpoints does not exceed 9000 points and that the dwell_time is also a multiple of the system clock frequency. 

I forgot how we reasoned that the calibration dwell time automatically is also a multiple of the system clock. Can you remember? On the other hand this was only true if the descanner period of 10us stays the same. That is why you added the extra check above, correct? However, in reality they are independent. I would suggest to also check here:

1. calibration dwell time is multiple of 25ns (system lock period)
2. The 9000 setpoints are not exceeded
3. calibration dwell time is >= min dwell time allowed on scanner (400ns)

Reply:
Reply on the first part of the first comment, yes indeed I forgot to upload that.
Reply on the second part of the first comment, indeed I think we should also implement that. Up to what frequencies do we want to restrict the system? Eventually you can have to little points for the descanner to still send a clear signal to the ASM.

Reply:
Currenlty the setpoints cannot exceed 9000 because of linespace. Which is we want to support more frequencies should be changed.
The calibration dwell time is per definition a multiple of the clock frequency because:
> Calibration_dwell_time = int((total_line_scan_time/9000) / scanner.clockPeriod.value)

The last division (and the int) explicitly says that the calibartion dwell time is a whole multiple of the scanner clockperiod.

Reply:
Reply on enumerated part second comment:

1. I think that is implied in the definition of the calibartion dwell time, see above
2. That is defined by the 9000, if we want to support more frequencies (which we should) we should code that in such a way that it is explicitly implies that it is impossible to exceed that 9000 or check it.
3. That should be implemented via allowing more calibartion frequencies, this way you run into this error earlies in the code which is nicer.

Reply:
> Reply on the second part of the first comment, indeed I think we should also implement that. Up to what frequencies do we want to restrict the system? Eventually you can have to little points for the descanner to still send a clear signal to the ASM.
Use the minimum and maximum dwell time defined for the range of the dwell time VA.

Reply:
> Reply on first part second comment:
> Currenlty the setpoints cannot exceed 9000 because of linespace. Which is we want to support more frequencies should be changed.
> The calibration dwell time is per definition a multiple of the clock frequency because:
> 
> > Calibration_dwell_time = int((total_line_scan_time/9000) / scanner.clockPeriod.value)
> 
> The last division (and the int) explicitly says that the calibartion dwell time is a whole multiple of the scanner clockperiod.

The calibration dwell time should be in first place without the division by the scanner clock period. That gives you the calibration dwell time in seconds. Only for uploading you have to provide it in system clock ticks. 
`calibration_dwell_time = total_line_scan_time/9000`
This value needs to be checked as described in step 1. to 3. 
It is not possible to just use that value and divide it by the clock period and then make an int out of it. That indeed forces the calibration dwell time to be a multiple of the system clock period, but the dwell time does not match the line scan time anymore, which you started your calculation with. 

You may check similar as in line 307:

```
remainder = int((total_line_scan_time/9000) % scanner.clockPeriod.value)
if remainder is not 0:
     reduce the number of setpoints (the 9000) 
```

With this you already cover point 1. and 2. Additionally, you need to check 3. If 3. is not fulfilled, you also need to adjust the ```

We can also briefly discuss offline. 
number of setpoints (9000). 
calibration_dwell_time < 400ns -> use less setpoints
calibration_dwell_time > 100us -> cannot happen as it is calculated based on the line scan time, and this itself is restricted by other parameters such as the maximum cell_size etc. 
```